base/pointer.jl

```diff
@@ -150,3 +150,24 @@ isless(x::Ptr, y::Ptr) = isless(UInt(x), UInt(y))
 +(x::Ptr, y::Integer) = oftype(x, (UInt(x) + (y % UInt) % UInt))
 -(x::Ptr, y::Integer) = oftype(x, (UInt(x) - (y % UInt) % UInt))
 +(x::Integer, y::Ptr) = y + x
+
+"""
+Temporarily protects an object from being garbage collected, even
+if it would otherwise be unreferenced.
+
+The last argument is the expression to preserve objects during.
+The previous arguments are the objects to preserve.
+"""
+macro gc_preserve(args...)
+    syms = args[1:end-1]
+    for x in syms
+        isa(x, Symbol) || error("Preserved variable must be a symbol")
+    end
+    s, r = gensym(), gensym()
+    esc(quote
+        $s = $(Expr(:gc_preserve, syms...))
+        $r = $(args[end])
```

Comment:
Is there a way to either catch (and raise error) or support control flow out of this block?

Reply:
Depends on what you mean. Control flow out of the block should work fine. Whether the thing will still be rooted then will depend on dominance in general.